server/common/app_config.py

```diff
@@ -19,74 +38,312 @@ def todict(self):
 
 
 class AppConfig(object):
-    def __init__(self, **kw):
-        super().__init__()
-
-        # app inputs
-        self.datapath = None
-        self.dataroot = None
-        self.title = ""
-        self.about = None
-        self.scripts = []
-        self.layout = []
-        self.max_category_items = 100
-        self.diffexp_lfc_cutoff = 0.01
-        self.disable_diffexp = False
-        self.enable_reembedding = False
-        self.anndata_backed = False
-
-        # The index page when in multi-dataset mode:
-        #   False or None:  this returns a 404 code
-        #   True:  loads a test index page, which links to the datasets that are available in the dataroot
-        #   string/URL:  redirect to this URL:  flask.redirect(config.multi_dataset_index)
-        self.multi_dataset_index = False
-
-        # A list of allowed matrix types.  If an empty list, then all matrix types are allowed
-        self.multi_dataset_allowed_matrix_type = []
-
-        # TODO these options may not apply to all datasets in the multi dataset.
-        # may need to invent a way to associate these config parameters with
-        # specific datasets.
-        self.obs_names = None
-        self.var_names = None
+    def __init__(self):
```

Comment:
Suggestion to prevent future mistakes / be more paranoid:  add a check to ensure that complete_config() is called before the config is used.

In other words:
1. add an instance value called `is_completed`.   
    * set to False in '__init__'
    * Set it to True in `complete_config()`
    * Set to False in `update()` and `update_from_config_file()`
2. Add a check for True (throw exception if False) in all of the 'get_*' methods

Reply:
Will do